### PR TITLE
gh-92734: Add indentation feature to reprlib.Repr

### PR DESCRIPTION
**Pull request by finefoot**
#92734

### CODE REVIEW

File: Doc/library/reprlib.rst [hunk 1]
```diff
@@ -123,6 +123,16 @@ which format specific object types.
    similar manner as :attr:`maxstring`.  The default is ``20``.
 
 
+.. attribute:: Repr.indent
+
+   If this attribute is set to ``None`` (default), no indentation of the output
```
Comment (zmievsa) [inline Doc/library/reprlib.rst]: These docs are a little confusing and have to be read multiple times to understand, even though the API itself is more or less intuitive.

**Review by zmievsa — CHANGES_REQUESTED** [review]
I haven't found any issues with the code or the tests. The docs are my only concern.

File: Doc/library/reprlib.rst (hunk 1 above)
Comment (finefoot) [reply Doc/library/reprlib.rst]: Thank you. I appreciate the review and feedback. Unfortunately, I'm not sure how to improve the docs paragraph. I've asked for help here: https://discuss.python.org/t/hint-or-advice-how-to-improve-docs-about-reprlib-repr-indent-feature/15747

File: Doc/library/reprlib.rst [hunk 2]
```diff
@@ -123,6 +123,16 @@ which format specific object types.
...
+   takes place. If it is set to either a string or an integer value, each
+   recursive representation will be on a new line, indented according to the
+   recursion depth. In that case, the attribute defines the string used for each
+   level of indentation. An integer value is equivalent to a string of space
+   characters of the same length.
```
Comment (CAM-Gerlach) [inline Doc/library/reprlib.rst]: Here's a revised version of the above section attempting to make the text more clear, and adding examples to illustrate the behavior in an easily-accessible way.

```suggestion
   If this attribute is set to ``None`` (the default), the output is formatted
   with no line breaks or indentation, like the standard :func:`repr`.
   For example:

   .. code-block:: pycon

      >>> example = [
              1, 'spam', {'a': 2, 'b': 'spam eggs', 'c': {3: 4.5, 6: []}}, 'ham']
      >>> print(reprlib.repr(example))
      [1, 'spam', {'a': 2, 'b': 'spam eggs', 'c': {3: 4.5, 6: []}}, 'ham']

   If :attr:`~Repr.indent` is set to a string, each recursion level
   is placed on its own line, indented by that string:

   .. code-block:: pycon

      >>> reprlib.aRepr.indent = '-->'
      >>> print(reprlib.repr(example))
      [
      -->1,
      -->'spam',
      -->{
      -->-->'a': 2,
      -->-->'b': 'spam eggs',
      -->-->'c': {
      -->-->-->3: 4.5,
      -->-->-->6: [],
      -->-->},
      -->},
      -->'ham',
      ]

   Setting :attr:`~Repr.indent` to a positive integer value behaves as if it
   was set to a string with that number of spaces:

   .. code-block:: pycon

      >>> reprlib.aRepr.indent = 4
      >>> print(reprlib.repr(example))
      [
          1,
          'spam',
          {
              'a': 2,
              'b': 'spam eggs',
              'c': {
                  3: 4.5,
                  6: [],
              },
          },
          'ham',
      ]
```

File: Lib/test/test_reprlib.py [hunk 3]
```diff
@@ -224,6 +224,62 @@ def test_unsortable(self):
         r(y)
         r(z)
 
+    def test_indent(self):
+        # Initialize two Repr instances: r1 without indentation, r2 with
```
Comment (CAM-Gerlach) [inline Lib/test/test_reprlib.py]: Shouldn't these each be separate subtests?

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
I provided a suggested revision of the documentation section, as you requested on Discourse.

Additionally, some quick testing of a few simple cases locally reveals several apparent issues and odd behaviors, which really should be handled properly (and captured in the tests, which currently are very limited):

* [x] Setting `indent = 0` makes any recursive objects silently not print anything, instead of being a natural extension of the positive integer case as it should be and I'd expect, with line breaks but no indents (or at the absolute minimum raising an error). Example:

  ```pycon
  >>> example = [1, 'spam', {'eggs': 1, 'ham': 2}]
  >>> reprlib.aRepr.indent = 0
  >>> print(reprlib.repr(example))
  []
  ```

  What I'd expect instead, of course, is

  ```python
  [
  1,
  'spam',
  {
  'eggs': 1,
  'ham': 2,
  },
  ]

* [x] Even more unexpectedly, the same happens for an empty string `''`, which from the perspective of a user is even more confusing, and likewise should work as one would expect

* [x] Furthermore, negative integers _also_ produce the same behavior, instead of either being an extension of the non-negative case (indenting less nested objects more to the left, with the indent level zero being the most deeply nested object—which I don't think would be useful enough to justify the implementation complexity) or raising an error at any point (which it probably should).

* [x] Also, passing `False` produces the same result, while `True` is equivalent to the integer `1`—presuming the former case is fixed to work like `0`, a seasoned Python user might anticipate and understand this, but I worry that it might confuse users who expect `False` to act like `None` and `True` to use some default/automatic value, e.g. 4 spaces. My suggestion would be to raise an explicit error in this edge case rather than trying to guess at the user's intent, though I'm not sure what's most idiomatic here.

* [x] Objects that clearly won't work for `indent`, like `[]`, `{"test": 1}`, `4.2` and arbitrary `object()`s produce two different possible `TypeError`s (one for objects that don't support `__mul__` with an `int`, and one for those that can but cannot be `__add__`ed with a `str`), neither of which are necessarily clear to the user as to the root cause, and only when actually used to indent. IMO, it would be better to catch this error and `raise from` a more helpful one, e.g. `Repr.indent must be int or str, not {type(indent)}`. You _could_ also use a `@property` to do validation on `indent` at set-time, but that's not very idiomatic, at least for the stdlib.

File: Doc/library/reprlib.rst (hunk 2 above)
Comment (zmievsa) [reply Doc/library/reprlib.rst]: I am not sure we should handle negative numbers. It seems like a bit too much magic.

File: Doc/library/reprlib.rst (hunk 1 above)
Comment (finefoot) [reply Doc/library/reprlib.rst:128]: @Ovsyanka83 Do the suggested changes below resolve your request for improved docs?

File: Doc/library/reprlib.rst (hunk 2 above)
Comment (finefoot) [reply Doc/library/reprlib.rst]: Looks good to me. Thanks a lot! :) I'm always in favor of having short code examples but the current docs didn't have any, so I kept my addition in the same style.

Comment (finefoot) [reply Doc/library/reprlib.rst:133]: I have added you to the repository in case you didn't have writing permissions, yet. Feel free to commit your changes.

Comment (finefoot) [reply Doc/library/reprlib.rst]: @CAM-Gerlach I have added you to the repository in case you didn't have writing permissions, yet. Feel free to commit your changes.

Comment (CAM-Gerlach) [reply Doc/library/reprlib.rst]: I can make the examples shorter if you like.

> I have added you to the repository in case you didn't have writing permissions, yet. Feel free to commit your changes.

Is there not a <kbd>Commit Suggestion</kbd> button below the suggestion above? You should be able to simply press that to apply my change directly; no need for any of that.

**Comment by finefoot** [comment]
Thank you very much for the extensive feedback.

> and captured in the tests, which currently are very limited

I agree. Do you have a suggestion in which way to add more cases? Because the best solution I can come up with, would be to simply add more explicit tests for sample cases and edge cases and compare the expected string to the return value. You can see that I tried to add a more "sophisticated" test, based on the idea that after removing whitespace and commas the output should be identical to the output produced with `Repr.indent = None`. But I guess that's not really helping much and I can just remove it again, in favor of more explicit tests?

> Setting indent = 0 makes any recursive objects silently not print anything

Good catch! :) I didn't think of this edge case. `Repr.indent = ""` (and subsequently `Repr.indent = 0`) should be fixed by https://github.com/python/cpython/pull/92735/commits/f8a2db1494d06d7ba534fdc5c989caed1964c22d. The reason was `-len(indent)` evaluating to `-0`, which is `0`, which obviously doesn't mean "cut zero characters from end of sequence" but "cut everything until character position 0". Instead, `None` works for zero-length indentation strings.

> IMO, it would be better to catch this error and `raise from` a more helpful one, e.g. `Repr.indent must be int or str, not {type(indent)}`. You _could_ also use a `@property` to do validation on `indent` at set-time, but that's not very idiomatic, at least for the stdlib.

The other attributes don't have any explicit error handling either. That's why I kept it out of the code, but I am in favor of adding it, due to exactly the situations you described. :+1:  If `@property` isn't idiomatic for the stdlib as you say, I've implemented simple LBYL error handling in https://github.com/python/cpython/pull/92735/commits/721357c7250063b2d3bba4cd73e623b935af4ee2. I noticed that you mentioned `raise from`, yet I don't think it's necessary (i.e. helpful) to catch the exception from an EAFP approach; and we'd have to LBYL due to disallowing negative integers anyway. Am I overlooking something?

> Furthermore, negative integers (...) indenting less nested objects more to the left, with the indent level zero being the most deeply nested object—which I don't think would be useful enough to justify the implementation complexity

I agree with this, too. Therefore, I wrote "non-negative int" in the `ValueError` message, above.

File: Doc/library/reprlib.rst (hunk 2 above)
Comment (finefoot) [reply Doc/library/reprlib.rst]: @CAM-Gerlach I thought it won't show you as the commit author if I commit via <kbd>Commit Suggestion</kbd>. So I added the `Co-authored-by:` tag, but apparently GitHub adds that anyway. :)

File: Lib/reprlib.py [hunk 4]
```diff
@@ -62,6 +63,23 @@ def repr1(self, x, level):
...
+        indent = self.indent
+        if not isinstance(indent, str):
+            if not isinstance(indent, int) or indent < 0:
+                raise ValueError(
+                    "Repr.indent must be None, str or non-negative int, not "
+                    + repr(indent)
+                )
+        if isinstance(indent, int):
+            indent *= ' '
+        sep = ',\n' + (self.maxlevel - level + 1) * indent
```
Comment (CAM-Gerlach) [inline Lib/reprlib.py]: I strongly recommend revising the error handling as follows. This raise the correct error class and message for the different cases (`TypeError` for passing an incompatible type, and `ValueError` for an inappropriate int value), use Pythonic EAFP, and `raise from` the actual error. And thanks to some cleanup and rationalization (using f-strings, avoiding an unnecessarily nested `if` statement, and a duplicate `isinstance` check) its only a couple lines longer.

```suggestion
        if isinstance(indent, int):
            if indent < 0:
                raise ValueError(
                    'Repr.indent cannot be negative int (was {indent!r})')
            indent *= ' '
        try:
            sep = ',\n' + (self.maxlevel - level + 1) * indent
        except TypeError as error:
            raise TypeError(
                f'Repr.indent must be of type None, str or int, not {type(indent)}'
            ) from error
```

**Comment by CAM-Gerlach** [comment]
> I agree. Do you have a suggestion in which way to add more cases?

To be honest, every other project I work on uses Pytest, where I'd just use the `@pytest.parameterize()` decorator for this, and I'm not terribly familiar with the idiomatic `unittest` approach—especially since the construct that I believe is most appropriate here, subtests, is the one main unittest feature that Pytest does not natively support.

That said, here's a stab at it, based on skimming over a few other examples in the CPython test suite (e.g. `test_codecs.py`). To summarize, you want methods for each major distinction in behavior (e.g. formatted output, raising an exception), and within each test method, you can use [unittest subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests), with one subtest per test case. iterating over a mapping of your inputs to expected outputs, with a key you can use to name the subtests.

Putting it all together, that would look something like this:

```python
import textwrap

...

    def test_valid_indent(self):
        test_obj = [1, 'spam', {'eggs': True, 'ham': []}]
        test_cases = {
            None: """[1, 'spam', {'eggs': True, 'ham': []}]""",
            4: ("""
                [
                    1,
                    'spam',
                    {
                        'eggs': True,
                        'ham': [],
                    },
                ]
                """),
            '-->': ("""
                [
                -->1,
                -->'spam',
                -->{
                -->-->'eggs': True,
                -->-->'ham': [],
                -->},
                ]
                """),
            # Etc.
        }

        for indent, expected_repr in test_cases.items():
            with self.subTest(indent=indent):
                repr_obj = Repr()
                repr_obj.indent = indent
                actual_repr = repr_obj.repr(test_obj).strip('\n')
                expected_repr = textwrap.dedent(expected_repr).strip('\n')
                self.assertEqual(actual_repr, expected_repr)

    def test_invalid_indent(self):
        test_obj = [1, 'spam', {'eggs': True, 'ham': []}]
        test_cases = {
            -2: (ValueError, '[Nn]egative|[Pp]ositive'),
            (4,): (TypeError, None),
            # Etc.
        }
        for indent, (expected_error, expected_msg) in test_cases.items():
            with self.subTest(indent=indent):
                repr_obj = Repr()
                repr_obj.indent = indent
                expected_msg = expected_msg or f'{indent!r}'
                with self.assertRaisesRegex(expected_error, expected_msg):
                    repr_obj.repr(test_obj)
```

> Good catch! :) I didn't think of this edge case. 

Its really helpful to get in the habit of doing so. Pretty much any time you're testing functions like this, even just as a quick manual sanity check, you should try common cases like `1`, `0`, `-1`, `True`, `False`, floats, empty strings, empty collections, arbitrary objects, etc. as appropriate for your use case, and make sure they behave how you expect (whether that's normal output or raising an appropriate error).

> Repr.indent = "" (and subsequently Repr.indent = 0) should be fixed by https://github.com/python/cpython/commit/f8a2db1494d06d7ba534fdc5c989caed1964c22d.

Indeed; LGTM in manual testing.

> I noticed that you mentioned raise from, yet I don't think it's necessary (i.e. helpful) to catch the exception from an EAFP approach; and we'd have to LBYL due to disallowing negative integers anyway. Am I overlooking something?

Why not? They are two very different cases that need to be handled separately, an invalid type (TypeError) and an invalid value (ValueError), and you already have a `if isinstance(indent, int)` block anyway for the invalid value (negative int) case. See above for a better approach.

File: Lib/reprlib.py (hunk 4 above)
Comment (finefoot) [reply Lib/reprlib.py]: Yes, I like the distinction between `TypeError` and `ValueError` :+1:

File: Lib/reprlib.py [hunk 5]
```diff
@@ -62,6 +63,27 @@ def repr1(self, x, level):
...
+                raise ValueError(
+                    f'Repr.indent cannot be negative int (was {indent!r})'
+                )
+            indent *= ' '
+        try:
+            sep = ',\n' + (self.maxlevel - level + 1) * indent
+        except TypeError as error:
+            raise TypeError(
+                'Repr.indent must be of type NoneType, str or int, '
+                f'not {type(indent)}'
```
Comment (CAM-Gerlach) [inline Lib/reprlib.py]: As `None` is an interpreter/language-reserved singleton, it is redundant, verbose and unidiomatic to refer to its type as `NoneType` instead of simply `None` in the context of a user-facing error message. This is also how its type is referred to in static type annotations and type checkers. Furthermore, I grepped all the `TypeError`s raised by the Python standard library that have the string `"None"` or `NoneType` somewhere in the first line of the exception text, and all of them referred to it as `None` alongside other types while none referred to it as `NoneType`.

To be shorter and more idiomatic while avoiding this issue entirely, we could slightly rewrite the message as follows:

```suggestion
                f'Repr.indent must be a str, int or None, not {type(indent)}'
```

Comment (finefoot) [reply Lib/reprlib.py]: Huh, fair enough, I guess. :) Thanks!

**Comment by finefoot** [comment]
> Also, passing `False` produces the same result, while `True` is equivalent to the integer `1`—presuming the former case is fixed to work like `0`, a seasoned Python user might anticipate and understand this, but I worry that it might confuse users who expect `False` to act like `None` and `True` to use some default/automatic value, e.g. 4 spaces. My suggestion would be to raise an explicit error in this edge case rather than trying to guess at the user's intent, though I'm not sure what's most idiomatic here.

I think we can reasonably skip this, in accordance with `json.dumps`, for example:

```
>>> print(json.dumps([[1]], indent=True))
[
 [
  1
 ]
]
>>> print(json.dumps([[1]], indent=False))
[
[
1
]
]
```

But I don't feel strongly about it. So if you disagree, I'm happy to open a thread and ask for opinions if there's a different idiomatic solution that would be best to implement. (For both `json.dumps` and `reprlib.Repr` in that case, I guess.)

> > I agree. Do you have a suggestion in which way to add more cases?
> 
> To be honest, every other project I work on uses Pytest, where I'd just use the `@pytest.parameterize()` decorator for this (...) That said, here's a stab at it, based on skimming over a few other examples in the CPython test suite

I apologize for the misunderstanding. I am aware of `unittest.TestCase.subTest` but simply didn't use it because I wanted to keep the indent tests in accordance with the preexisting tests in `test_reprlib.py`. I was under the impression that document coherence is more important. I am fine with using subtests. :+1: 

My question was meant more like this: Theoretically, we could reverse the indentation process and test if the result matches the result that regular `Repr.indent = None` produces. The rough idea was like this:

```python
# Results are the same after removing whitespace and commas
y = [1, [2, "foo", b"bar", {"a": 1, "b": "abc def ghi", "c": {1: 2, 3: 4, 5: [], 6: {}}}], 3]
self.assertNotEqual(r1.repr(y), r2.repr(y))
self.assertEqual(
    "".join(r1.repr(y).replace(",", "").split()),
    "".join(r2.repr(y).replace(",", "").split())
)
```

But, like I said before, I don't think it's worth it. Explicit tests of example cases should be fine? I was simply inquiring if you agreed with this or if you think I should work on the "indentation reversal test". :)

Anyway, I've taken your suggestions and added some tests in https://github.com/python/cpython/pull/92735/commits/1d652a5be697c74a014644cff1345cf434450e5b. I also split it into `test_valid_indent` and `test_invalid_indent`, like you proposed.

**Comment by CAM-Gerlach** [comment]
> I think we can reasonably skip this, in accordance with json.dumps, for example:

I think you're right; its consistent with the behavior elsewhere and its not _that_ big of a deal. You should, however, include `True` and `False` in your `indent` value test cases, which otherwise look fairly complete.

> I apologize for the misunderstanding. 

No need to apologize!

> Anyway, I've taken your suggestions and added some tests in https://github.com/python/cpython/commit/1d652a5be697c74a014644cff1345cf434450e5b. I also split it into test_valid_indent and test_invalid_indent, like you proposed.

All looks good; aside from including `True` and `False` as test cases for `indent`, to avoid readability impacts due to the space it consumes, I might might your cases cases to a module-level variable, and you could also use somewhat smaller test objects, though its fine if others don't complain. You could also consider using nested subtests (object ---> settings). Other than that, LGTM at least from my end.

**Comment by finefoot** [comment]
> You should, however, include `True` and `False` in your `indent` value test cases,

:+1: Tests added in https://github.com/python/cpython/pull/92735/commits/7cc9fc839a41031d322aea989d7595f2fee6851e. I assume, the rationale behind testing this undefined behavior - since the docs don't mention `bool` for `Repr.indent` - is less about verifying that `indent = True` produces the same as `indent = 1`, but more about verifying that no error happens if a user unknowingly uses said value producing undefined behavior?

> to avoid readability impacts due to the space it consumes, I might might your cases cases to a module-level variable, and you could also use somewhat smaller test objects, though its fine if others don't complain

The example in `test_pprint.QueryTestCase.test_set_of_sets_reprs` is fairly big, too, for example. I guess that means, it's okay?

> You could also consider using nested subtests (object ---> settings)

That's a very good suggestion, thank you. :) Implemented in https://github.com/python/cpython/pull/92735/commits/61a6b5aed6fd6d1edaa70ee64451ce8d02ad98cf.

**Comment by CAM-Gerlach** [comment]
> +1 Tests added in https://github.com/python/cpython/commit/7cc9fc839a41031d322aea989d7595f2fee6851e. I assume, the rationale behind testing this undefined behavior - since the docs don't mention bool for Repr.indent - is less about verifying that indent = True produces the same as indent = 1, but more about verifying that no error happens if a user unknowingly uses said value producing undefined behavior?

Looks good. It has a dual purpose, in my mind—it both tests the current behavior, and serves as de-facto CPython-developer-facing, machine-checkable verification of it; if it is changed, intentionally or not, then the test would provide clear indication of that and would ensure the tests are updated to cover the new behavior instead.

> The example in test_pprint.QueryTestCase.test_set_of_sets_reprs is fairly big, too, for example. I guess that means, it's okay?

I don't have a problem with the size, though personally I'd separate the test data from the test code to make it cleaner to follow (in Pytest, I'd made it a fixture or at least a module-level constant). I was mostly concerned about any objections from others with more authority about it, since I'm not a core dev myself. But its ultimately up to you and whatever core dev does the final review and merge here. Otherwise, LGTM!

**Review by CAM-Gerlach — APPROVED** [review]
Thanks for all your hard work and responsiveness to feedback, @finefoot . LGTM now and looking forward to see this merged; this should be ready for core dev review now (with one potential concern about mixing test data and code outstanding, though that can be left to a core dev for final judgment).

**Comment by CAM-Gerlach** [comment]
Tagging @rhettinger for review, since there's no one listed in the [Experts Index](https://devguide.python.org/experts/#stdlib) or CODEOWNERS and he seems to be de-facto responsible for this module given his code contributions and his review of the most recent significant PR affecting it.

**Comment by finefoot** [comment]
> with one potential concern about mixing test data and code outstanding, though that can be left to a core dev for final judgment

Alright, makes sense. Let's see. :)

@CAM-Gerlach Thanks for your patience, explanations and helpful feedback. I learned a lot :+1:

**Comment by CAM-Gerlach** [comment]
Thank you to, so did I! I'm pretty new to most of this stuff as well and have only just started contributing to CPython proper in the past couple weeks, and pretty much all of that has been docs stuff so far. I'm also not even a "real" software developer, I just play one on the internet—in real life, I'm a NASA satellite scientist, heh.

**Comment by finefoot** [comment]
> If your pull request has not received any notice from reviewers (i.e., no comment made) after one month, first “ping” the issue on the [issue tracker](https://github.com/python/cpython/issues) to remind the nosy list that the pull request needs a review.

Source: https://devguide.python.org/pullrequest/

**Comment by finefoot** [comment]
> If your pull request has not received any notice from reviewers (i.e., no comment made) after one month, first “ping” the issue on the [issue tracker](https://github.com/python/cpython/issues) to remind the nosy list that the pull request needs a review.

Source: https://devguide.python.org/pullrequest/

Ping :wink:

**Comment by CAM-Gerlach** [comment]
Thanks, and sorry for the delay; I've reached out to the core devs to take a look at this...

File: Doc/library/reprlib.rst [hunk 6]
```diff
@@ -123,6 +123,62 @@ which format specific object types.
    similar manner as :attr:`maxstring`.  The default is ``20``.
 
 
+.. attribute:: Repr.indent
+
+   If this attribute is set to ``None`` (the default), the output is formatted
+   with no line breaks or indentation, like the standard :func:`repr`.
```
Comment (ambv) [inline Doc/library/reprlib.rst:148]: I think a more sensible default value would be just `""` (an empty string). In other words, *only support* the string indentation variant. This will let users say `4 * " "` if they want for spaces, and `"\t"` if they want tabs. The implementation becomes simpler then too as you only have to deal with strings.

In this case we wouldn't be handling the case where you explode arguments but don't indent. I think this is fair game as the attribute is called `indent` anyway.

File: Doc/library/reprlib.rst [hunk 7]
```diff
@@ -123,6 +123,62 @@ which format specific object types.
...
+      -->},
+      -->'ham',
+      ]
+
+   Setting :attr:`~Repr.indent` to a positive integer value behaves as if it
+   was set to a string with that number of spaces:
+
+   .. code-block:: pycon
+
+      >>> reprlib.aRepr.indent = 4
```
Comment (ambv) [inline Doc/library/reprlib.rst]: The documentation doesn't currently recommend monkey-patching attributes on globally used objects. We shouldn't recommend that here either. The danger here is that it's a global state change, which might affect other parts of your process. Like, some third-party library you're using might be using reprlib for a purpose where an indent change (especially non-whitespace indentation) might break it.

Instead, I'd create a new `Repr()` instance in this example and modify `indent` of that object.

File: Misc/NEWS.d/next/Core and Builtins/2022-05-12-15-19-00.gh-issue-92734.d0wjDt.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+Add an indentation feature to ``reprlib.Repr``
```
Comment (ambv) [inline Misc/NEWS.d/next/Core and Builtins/2022-05-12-15-19-00.gh-issue-92734.d0wjDt.rst]: This NEWS entry is misplaced (it's not "Core", it's "Library") and it's somewhat terse. How about something like: 
```suggestion
Allow multi-element reprs emitted by :mod:`reprlib` to be pretty-printed using configurable indentation.
```

**Review by ambv — COMMENTED** [review]
In general I find it icky that `indent` is a field here and not an argument to the `repr()` method. However, since the rest of the functionality in `reprlib` provides a precedent for this design, I think we can live with that.

That being said, I have a few comments around the current implementation.

File: Doc/library/reprlib.rst (hunk 6 above)
Comment (finefoot) [reply Doc/library/reprlib.rst:129]: Thanks for the suggestion. The original idea was to have `None` and integer values as the the basic types for the `indent` argument because it seems appropriate to keep the basics of the `indent` attribute similar to the `indent` arguments of [`json.dump`](https://docs.python.org/3/library/json.html#json.dump) as well as [`pprint.PrettyPrinter`](https://docs.python.org/3/library/pprint.html#pprint.PrettyPrinter) which use `None` and integer values, respectively. On top of that, allowing string values is a useful extension for certain cases. (`"\t"` being an obvious one.) Taking all this into consideration, I don't feel like having `reprlib.Repr.indent` support only string values would be the right choice.

File: Misc/NEWS.d/next/Core and Builtins/2022-05-12-15-19-00.gh-issue-92734.d0wjDt.rst (hunk 8 above)
Comment (finefoot) [reply Misc/NEWS.d/next/Core and Builtins/2022-05-12-15-19-00.gh-issue-92734.d0wjDt.rst]: My bad. Will fix the filename. Thanks for the improved phrasing. :+1:

Comment (finefoot) [reply Misc/NEWS.d/next/Core and Builtins/2022-05-12-15-19-00.gh-issue-92734.d0wjDt.rst]: File has been moved f7e91b6cb5deec8bf637707b000c4f30e890c5bb

File: Doc/library/reprlib.rst (hunk 7 above)
Comment (finefoot) [reply Doc/library/reprlib.rst]: Sounds good. I have adjusted the docs examples to use a fresh instance of `reprlib.Repr` instead of modifying the global `reprlib.aRepr` object in df43a670847aff53a25c94639f873c28d59cc702. I chose the variable name `aRepr` for the "example instance" to stay consistent with the other example code further down on the `reprlib` docs page.

File: Doc/library/reprlib.rst (hunk 6 above)
Comment (finefoot) [reply Doc/library/reprlib.rst:148]: Thanks for the suggestion. The original idea was to have `None` and integer values as the the basic types for `Repr.indent` because it seems appropriate to keep the basics of the `indent` attribute similar to the `indent` arguments of [`json.dump`](https://docs.python.org/3/library/json.html#json.dump) as well as [`pprint.PrettyPrinter`](https://docs.python.org/3/library/pprint.html#pprint.PrettyPrinter) which use `None` and integer values, respectively. On top of that, allowing string values is a useful extension for certain cases. (`"\t"` being an obvious one.) Taking this into consideration, I think I don't feel like having `reprlib.Repr.indent` support only string values would be the right choice, but could lead to confusion.

**Comment by finefoot** [comment]
> Thanks, and sorry for the delay; I've reached out to the core devs to take a look at this...

@CAM-Gerlach No worries. :) Out of curiosity: Did you just use one of the mailing lists? Is that something I should have done instead? It's mentioned on the Python Developer's Guide, too. But only as the next point following the ping here on the GitHub tracker.

**Comment by finefoot** [comment]
I guess this would make sense to add to the docs once it's getting close to getting merged and the Python version is foreseeable?

- [ ] Add `   .. versionadded:: 3.xyz` to docs

It doesn't get added automatically, right?

**Comment by CAM-Gerlach** [comment]
> I guess this would make sense to add to the docs once it's getting close to getting merged and the Python version is foreseeable?

    Add .. versionadded:: 3.xyz to docs

It doesn't get added automatically, right?

Oops, I completely forgot the mention that—great catch. No, its not automatic, and as this is essentially certain to go in Python 3.12 now, we should just go ahead and added it. I'll make a suggestion to that effect.

File: Doc/library/reprlib.rst [hunk 9]
```diff
@@ -123,6 +123,64 @@ which format specific object types.
...
+              'a': 2,
+              'b': 'spam eggs',
+              'c': {
+                  3: 4.5,
+                  6: [],
+              },
+          },
+          'ham',
+      ]
+
```
Comment (CAM-Gerlach) [inline Doc/library/reprlib.rst:201]:
```suggestion
   .. versionadded:: 3.12
```

Add the version added.

File: Doc/library/reprlib.rst (hunk 6 above)
Comment (CAM-Gerlach) [reply Doc/library/reprlib.rst:148]: While such an approach is seemingly alluring, upon close inspection this doesn't appear to actually allow simplifying the code much, as a `""` value would is still just as much a special case that would need to be handled specially, just as we do now (unless we change the default to explode the repr, which is almost certainly not what we want)—just with less clarity, since `None` more obviously is a special default case than an empty string. The only block that could be obviously eliminated is 

```python
        if isinstance(indent, int):
            if indent < 0:
                raise ValueError(
                    f'Repr.indent cannot be negative int (was {indent!r})'
                )
            indent *= ' '
```

which is by a fair margin most straightforward and easy to understand code in the added function, relative to the much more complex string-munging bits, and that is only due to not accepting `int` arguments; changing the default value would not enable any further simplification beyond that.

By contrast, the behavior is potentially more confusing and less obvious and useful to users, since `""` is magically handled differently than any other string and breaks the natural expectation arising from the pattern when passing `"    "`, `"  "`, `" "` or `" " * 4`, `" " * 2`, `" " * 1`, and that leaves no way to set an indent level of zero but still explode a string—I'm not sure how else this would be done.

Given all that, as well as the points raised by @finefoot above and the fact that implementation is complete and has fairly extensive unit and manual tests for a wide array of cases, I'd suggest retaining the design as it stands now. If supporting integer as well as string arguments is a hard blocker, then at worst that could be removed while retaining `None` as a default value, allowing `""` to be passed, as changing that would not make the implementation any more simple while artificially limiting functionality and quite possibly confusing users).

File: Doc/library/reprlib.rst (hunk 9 above)
Comment (finefoot) [reply Doc/library/reprlib.rst:201]: Thanks :)

**Comment by finefoot** [comment]
@ambv Friendly ping: I have implemented your suggested changes regarding removing the monkey patching from the docs example as well as the description in NEWS - thanks again :) - and we replied to your other comment to explain our reasoning.

**Comment by CAM-Gerlach** [comment]
FYI, right now things are very intense due to the planned release of 3.11rc1, so if @ambv doesn't respond in a week or two, ping again and I can ping him or another core dev if there still isn't a response.

**Comment by finefoot** [comment]
Thanks for letting me know. That's very nice of you to keep me updated on the cause of delays. It's absolutely not necessary, though. ;) I know people are busy, I know it's going to take time to review pull requests. No worries here. I just wanted to ping again after a month of inactivity like it says in the Python dev guide.
